In [1]:
import requests
import pandas as pd

Part 1: Data Collection 

In [2]:
with open("Section4-5_Findings_MohammedAli.txt", "w") as f:
    f.write("Introduction\n")
    f.write("I collected 100 rows of data from the open meteo api for nairobi kenya from 1st may 2026 to 24th august 2026.\n"
    "The data was collected in json format and converted to a pandas dataframe.\n"
    "The dataframe was then cleaned by renaming the columns, converting the date column to datetime format, and extracting the month from the date column.\n"
    "The cleaned dataframe was then saved to a csv file named nairobi_weather_cleaned.csv.\n"
    "The dataframe was then analyzed by calculating the value counts and mode for the month column, calculating the skewness for the numeric columns, calculating the correlation matrix for the numeric columns,\n "
    "and calculating the average temp_max_c by month using groupby.")

In [3]:
# i have chosen the open meteo api to fetch data for nairobi kenya from 1st may 2026 to 24th august 2026.
url = "https://archive-api.open-meteo.com/v1/archive"
params = {
    "latitude":-1.2864 ,
    "longitude":36.8172,
    "start_date": "2026-05-01",
    "end_date": "2026-08-24",  
    "daily": "temperature_2m_max,temperature_2m_min,precipitation_sum,windspeed_10m_max",
    "timezone": "auto",
}

In [4]:
# fetching 100 rows and saving to a pandas dataframe
response = requests.get(url, params=params)
data = response.json()['daily']
df = pd.DataFrame(data)
df

,time,temperature_2m_max,temperature_2m_min,precipitation_sum,windspeed_10m_max
0,2026-05-01,23.0,16.0,10.8,8.1
1,2026-05-02,23.2,16.1,5.3,11.0
2,2026-05-03,23.9,14.4,2.3,12.2
3,2026-05-04,23.8,15.0,2.3,9.4
4,2026-05-05,23.8,16.4,1.8,10.4
...,...,...,...,...,...
111,2026-08-20,25.1,13.1,0.2,17.6
112,2026-08-21,25.0,11.1,0.0,9.5
113,2026-08-22,27.5,12.0,0.0,12.0
114,2026-08-23,27.3,16.1,2.8,11.4


In [5]:
# renaming the columns to more meaningful names 
df = df.rename(columns={
    "time": "date",
    "temperature_2m_max": "temp_max_c",
    "temperature_2m_min": "temp_min_c",
    "precipitation_sum": "precipitation_mm",
    "windspeed_10m_max": "windspeed_max_kmh",
})

In [6]:
# printing the first 5 rows of the dataframe
print(f"Rows fetched: {len(df)}")
print(df.head())

Rows fetched: 116
         date  temp_max_c  temp_min_c  precipitation_mm  windspeed_max_kmh
0  2026-05-01        23.0        16.0              10.8                8.1
1  2026-05-02        23.2        16.1               5.3               11.0
2  2026-05-03        23.9        14.4               2.3               12.2
3  2026-05-04        23.8        15.0               2.3                9.4
4  2026-05-05        23.8        16.4               1.8               10.4


In [7]:
# saving the cleaned dataframe to a csv file
df.to_csv("nairobi_weather.csv", index=False)
print("Saved to nairobi_weather.csv")

Saved to nairobi_weather.csv


In [8]:
# checking for initial shape of the dataframe before cleaning
initial_shape = df.shape
print(f"Initial shape: {initial_shape}")

Initial shape: (116, 5)


In [9]:
# checking the data types of the columns in the dataframe
df.dtypes

date                  object
temp_max_c           float64
temp_min_c           float64
precipitation_mm     float64
windspeed_max_kmh    float64
dtype: object

What was the API endpoint you used?
v1/archive


How many records did you fetch? How many columns?
i got 116 rows of data and 5 columns


Did the API have any rate limits or auth requirements?
no, the open meteo does not have authorization reirement since it is free for non commercial users and has a rate limit of 10000 calls a day


What were the column data types initially?
the date is an object(date)
temp_max_c is a float
temp_min_c is a float
precipitation_mm is a float
windspeed_max_kmh is a float

Part 2: Data Cleaning 

In [10]:
# reading the nairobi_weather csv file
df = pd.read_csv("nairobi_weather.csv")

In [11]:
# using the info() method to get a concise summary of the dataframe, including the number of non-null values and data types for each column
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 116 entries, 0 to 115
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   date               116 non-null    object 
 1   temp_max_c         116 non-null    float64
 2   temp_min_c         116 non-null    float64
 3   precipitation_mm   116 non-null    float64
 4   windspeed_max_kmh  116 non-null    float64
dtypes: float64(4), object(1)
memory usage: 4.7+ KB


In [12]:
# using the describe() method to get a statistical summary of the dataframe, including count, mean, std, min, 25%, 50%, 75%, and max for each numeric column
df.describe()

,temp_max_c,temp_min_c,precipitation_mm,windspeed_max_kmh
count,116.000000,116.000000,116.000000,116.000000
mean,24.186207,14.082759,1.018966,12.127586
std,1.458462,1.527332,1.769250,2.446278
min,20.600000,10.600000,0.000000,5.800000
25%,23.300000,12.800000,0.100000,10.300000
50%,24.250000,14.150000,0.400000,12.000000
75%,25.000000,15.500000,0.950000,13.500000
max,27.900000,16.700000,10.800000,18.800000


In [14]:
# viewing the first 10 rows of the dataframe
df.head(10)

,date,temp_max_c,temp_min_c,precipitation_mm,windspeed_max_kmh
0,2026-05-01,23.0,16.0,10.8,8.1
1,2026-05-02,23.2,16.1,5.3,11.0
2,2026-05-03,23.9,14.4,2.3,12.2
3,2026-05-04,23.8,15.0,2.3,9.4
4,2026-05-05,23.8,16.4,1.8,10.4
5,2026-05-06,24.8,16.4,2.5,8.8
6,2026-05-07,24.3,16.1,1.2,9.4
7,2026-05-08,25.0,16.6,1.7,9.5
8,2026-05-09,23.5,16.6,8.4,9.3
9,2026-05-10,23.8,15.3,0.7,11.9


In [15]:
# checking for missing values in the dataframe
print(df.isnull().sum())

date                 0
temp_max_c           0
temp_min_c           0
precipitation_mm     0
windspeed_max_kmh    0
dtype: int64


In [16]:
# checking for missing values in the dataframe and calculating the percentage of missing values for each column
print((df.isnull().sum() / len(df) * 100).round(2))

date                 0.0
temp_max_c           0.0
temp_min_c           0.0
precipitation_mm     0.0
windspeed_max_kmh    0.0
dtype: float64


In [17]:
# checking for duplicate rows in the dataframe
print(f"Duplicate rows: {df.duplicated().sum()}")

Duplicate rows: 0


In [19]:
# Fixing the  data types (convert strings to dates/numbers as needed)
df['date'] = pd.to_datetime(df['date'])
df['temp_max_c'] = pd.to_numeric(df['temp_max_c'])
df['temp_min_c'] = pd.to_numeric(df['temp_min_c'])
df['precipitation_mm'] = pd.to_numeric(df['precipitation_mm'])
df['windspeed_max_kmh'] = pd.to_numeric(df['windspeed_max_kmh'])
print(df.dtypes)
df.head(10)

date                 datetime64[ns]
temp_max_c                  float64
temp_min_c                  float64
precipitation_mm            float64
windspeed_max_kmh           float64
dtype: object


,date,temp_max_c,temp_min_c,precipitation_mm,windspeed_max_kmh
0,2026-05-01,23.0,16.0,10.8,8.1
1,2026-05-02,23.2,16.1,5.3,11.0
2,2026-05-03,23.9,14.4,2.3,12.2
3,2026-05-04,23.8,15.0,2.3,9.4
4,2026-05-05,23.8,16.4,1.8,10.4
5,2026-05-06,24.8,16.4,2.5,8.8
6,2026-05-07,24.3,16.1,1.2,9.4
7,2026-05-08,25.0,16.6,1.7,9.5
8,2026-05-09,23.5,16.6,8.4,9.3
9,2026-05-10,23.8,15.3,0.7,11.9


In [20]:
# Remove or flag outliers (document your decision)
def flag_outliers(col):
    q1, q3 = df[col].quantile(0.25), df[col].quantile(0.75)
    iqr = q3 - q1
    lower, upper = q1 - 1.5 * iqr, q3 + 1.5 * iqr
    print(f"{col}: lower={lower}, upper={upper}")
    return (df[col] < lower) | (df[col] > upper)

for col in ["temp_max_c", "temp_min_c", "precipitation_mm", "windspeed_max_kmh"]:
    flag_col = f"{col}_outlier"
    df[flag_col] = flag_outliers(col)
    print(f"Outliers flagged in {col}: {df[flag_col].sum()}")

temp_max_c: lower=20.75, upper=27.549999999999997
Outliers flagged in temp_max_c: 2
temp_min_c: lower=8.750000000000002, upper=19.549999999999997
Outliers flagged in temp_min_c: 0
precipitation_mm: lower=-1.175, upper=2.225
Outliers flagged in precipitation_mm: 18
windspeed_max_kmh: lower=5.500000000000002, upper=18.299999999999997
Outliers flagged in windspeed_max_kmh: 1


<!-- I used the IQR method for outlier detection.
Precipitation had the most outliers (18), mainly because rainfall varies a lot, with most days being dry and a few having heavy rain
These were kept since they represent real weather events.
Temp_max_c: 2 outliers — a few unusually hot days.
Windspeed_max_kmh: 1 outlier — an unusually windy day.
Temp_min_c: No outliers — nighttime temperatures were fairly stable.
All outliers were flagged but not removed -->

In [21]:
# checking for duplicate rows in the dataframe
duplicate_rows=df.duplicated().sum()
print(f"Duplicate rows: {duplicate_rows}")

Duplicate rows: 0


In [22]:
# checking for the shape of the dataframe after cleaning
current_shape = df.shape
print(f"Current shape: {current_shape}")

Current shape: (116, 9)


In [23]:
# documenting my findings
with open("Section4-5_Findings_MohammedAli.txt", "a") as f:
    f.write("\nFindings from Nairobi Weather Data Analysis\n")
    f.write("="*50 + "\n")
    f.write(f"Total rows fetched: {len(df)}\n")
    f.write(f"Missing values:\n{df.isnull().sum()}\n")
    f.write(f"Percentage of missing values:\n{(df.isnull().sum() / len(df) * 100).round(2)}\n")
    f.write(f"Duplicate rows: {df.duplicated().sum()}\n")
    for col in ["temp_max_c", "temp_min_c", "precipitation_mm", "windspeed_max_kmh"]:
        flag_col = f"{col}_outlier"
        f.write(f"Outliers flagged in {col}: {df[flag_col].sum()}\n")
    f.write("="*100 + "\n")
    f.write("I used the IQR method for outlier detection.\n")
    f.write("I kept since they represent real weather events.\n")
    f.write("All outliers were flagged but not removed.\n")
    f.write("="*100 + "\n")
    f.write(f"Initial shape: {initial_shape}\n")
    f.write(f"Current shape: {current_shape}\n")
    
    


In [24]:
df.to_csv("nairobi_weather_cleaned.csv", index=False)

How many missing values did you find? In which columns?
i did not find any missing values in the dataset i have

Did you drop rows, fill values, or keep them? Why?
i did not drop any rows

Were there duplicates? How did you handle them?
there were no duplicates

Did any data type conversions surprise you?
the float columns remained while the date column changed to datetime

Part 3: Exploratory Data Analysis

In [25]:
# reading the cleaned nairobi_weather csv file
df = pd.read_csv("nairobi_weather_cleaned.csv")
df["date"] = pd.to_datetime(df["date"])

In [26]:
# creating a categorical column since the raw dataset is all numeric.'month' gives something meaningful to group weather patterns by.
df["month"] = df["date"].dt.month_name()
df

,date,temp_max_c,temp_min_c,precipitation_mm,windspeed_max_kmh,temp_max_c_outlier,temp_min_c_outlier,precipitation_mm_outlier,windspeed_max_kmh_outlier,month
0,2026-05-01,23.0,16.0,10.8,8.1,False,False,True,False,May
1,2026-05-02,23.2,16.1,5.3,11.0,False,False,True,False,May
2,2026-05-03,23.9,14.4,2.3,12.2,False,False,True,False,May
3,2026-05-04,23.8,15.0,2.3,9.4,False,False,True,False,May
4,2026-05-05,23.8,16.4,1.8,10.4,False,False,False,False,May
...,...,...,...,...,...,...,...,...,...,...
111,2026-08-20,25.1,13.1,0.2,17.6,False,False,False,False,August
112,2026-08-21,25.0,11.1,0.0,9.5,False,False,False,False,August
113,2026-08-22,27.5,12.0,0.0,12.0,False,False,False,False,August
114,2026-08-23,27.3,16.1,2.8,11.4,False,False,True,False,August


In [28]:
# creating a list of numeric columns for further analysis
numeric_cols = ["temp_max_c", "temp_min_c", "precipitation_mm", "windspeed_max_kmh"]
print(numeric_cols)

['temp_max_c', 'temp_min_c', 'precipitation_mm', 'windspeed_max_kmh']


In [29]:
# calculating the mean, median, standard deviation, minimum, and maximum for each numeric column and rounding to 2 decimal places
summary = df[numeric_cols].agg(["mean", "median", "std", "min", "max"]).round(2)
print(summary)

        temp_max_c  temp_min_c  precipitation_mm  windspeed_max_kmh
mean         24.19       14.08              1.02              12.13
median       24.25       14.15              0.40              12.00
std           1.46        1.53              1.77               2.45
min          20.60       10.60              0.00               5.80
max          27.90       16.70             10.80              18.80


In [30]:
# finding value counts for 'month' 
print(df["month"].value_counts())

month
May       31
July      31
June      30
August    24
Name: count, dtype: int64


In [31]:
#  finding mode for 'month'
print(df["month"].mode()[0])

July


In [32]:
# calculating skewness for each numeric column and rounding to 2 decimal places
print(df[numeric_cols].skew().round(2))

temp_max_c           0.02
temp_min_c          -0.05
precipitation_mm     3.25
windspeed_max_kmh    0.43
dtype: float64


In [33]:
# calculating the correlation matrix for the numeric columns and rounding to 2 decimal places
print(df[numeric_cols].corr().round(2))

                   temp_max_c  temp_min_c  precipitation_mm  windspeed_max_kmh
temp_max_c               1.00       -0.16             -0.24               0.36
temp_min_c              -0.16        1.00              0.54              -0.23
precipitation_mm        -0.24        0.54              1.00              -0.29
windspeed_max_kmh        0.36       -0.23             -0.29               1.00


In [34]:
#  calculating the average temp_max_c by month using groupby and rounding to 2 decimal places, then sorting in descending order
print(df.groupby("month")["temp_max_c"].mean().round(2).sort_values(ascending=False))

month
August    24.70
July      24.23
May       24.10
June      23.82
Name: temp_max_c, dtype: float64


In [35]:
# documenting my findings
with open("Section4-5_Findings_MohammedAli.txt", "a") as f:
    f.write("="*50 + "\n")
    f.write("\nSummary Statistics\n")
    f.write(summary.to_string())
    f.write("\n\n=== Categorical: Value Counts for 'month' ===\n")
    f.write(df["month"].value_counts().to_string())
    f.write("\n\n=== Categorical: Mode for 'month' ===\n")
    f.write(str(df["month"].mode()[0]))
    f.write("\n\n=== Skewness ===\n")
    f.write(df[numeric_cols].skew().round(2).to_string())
    f.write("\n\n=== Correlation Matrix (numeric pairs) ===\n")
    f.write(df[numeric_cols].corr().round(2).to_string())
    f.write("\n\n=== Groupby: Average temp_max_c by month ===\n")
    f.write(df.groupby("month")["temp_max_c"].mean().round(2).sort_values(ascending=False).to_string())
    f.write("\n" + "="*50 + "\n")
    
    

What surprised you?
i thought rainy(days with high precipitation) had cooler nights,the reverse is what i observerd

What patterns did you spot?
temperature highs remains similar around mid year


Any outliers or anomalies worth flagging?
no any new outliers/anomalies

Which column has the most missing data?
there are no columns with missing data in my dataset


What is the correlation between your strongest pair of variables?
i have observed that my strongest pair is between the precipitation_mm and the temp_min_c is a moderate positive corelation(0.54) which could mean that days with high precipitation_mm could have warmer nights


What's the most common value in your categorical column?
i picked the month of july since i was looking at weather patterns for the past 4 months and july has the most days compared to june


What's one unexpected pattern you found?
i found that on days when it rained more than usual, the night was often warmer than usual too because the pattern was a moderate positive correlation (r = 0.54) between precipitation_mm and temp_min_c

In [36]:
# conclusion
with open("Section4-5_Findings_MohammedAli.txt", "a") as f:
    f.write("\n"+"="*50 + "\n")
    f.write("\nConclusion\n")
    f.write("The next steps for this analysis includes visualizing the data using plots to better understand the trends and patterns in the weather data.")